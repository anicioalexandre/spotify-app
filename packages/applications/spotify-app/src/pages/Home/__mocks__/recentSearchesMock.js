const recentSearchesMock = {
  last_results: [
    {
      trackId: '3BQHpFgAp4l80e1XslIjNI',
      albumId: '0PT5m6hwPRrpBwIHVnvbFX',
      image: 'https://i.scdn.co/image/ab67616d00001e02e3e3b64cea45265469d4cafa',
      artistName: 'The Beatles',
      name: 'Yesterday - Remastered 2009'
    },
    {
      albumId: '1vANZV20H5B4Fk6yf7Ot9a',
      image: 'https://i.scdn.co/image/ab67616d00001e0255612ece447bec5d62c68375',
      name: 'Beatles For Sale (Remastered)',
      artistName: 'The Beatles'
    },
    {
      albumId: '2zeWAYgh0WfMhiHwvUV15b',
      image: 'https://i.scdn.co/image/ab67616d00001e029f08c2a93ef090a1e1e6bf0c',
      name: 'NÃ³ na Orelha',
      artistName: 'Criolo'
    }
  ]
}

module.exports = { recentSearchesMock }
