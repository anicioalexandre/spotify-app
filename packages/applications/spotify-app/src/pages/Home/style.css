.home-container {
  display: flex;
  flex-direction: column;
  gap: 50px;
}

.home-container a {
  text-decoration: none;
  transition: opacity 0.35s ease;
}

.home-container a:hover {
  opacity: var(--hover-opacity);
}
