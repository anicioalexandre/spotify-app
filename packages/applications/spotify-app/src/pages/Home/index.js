import React, { useEffect, useState } from 'react'
import { Link } from 'react-router-dom'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import 'ds-image-label'
import 'ds-fieldset'
import 'ds-button'

import './style.css'
import { LABELS } from './constants'
import SearchInput from '../../components/SearchInput'
import { clearLocalStorage, getLocalStorage } from '../../utils/localStorage'
import { setClickedTrack } from '../../redux/modules/track'
import EmptyState from '../../components/EmptyState'

const Home = ({ setClickedTrackAction }) => {
  const [recentSearches, setRecentSearches] = useState([])

  useEffect(() => {
    setRecentSearches(getLocalStorage())
  }, [])
  return (
    <div className="home-container">
      <SearchInput />
      <ds-fieldset label={LABELS.recentSearch}>
        {!recentSearches?.length ? (
          <EmptyState customLabel={LABELS.emptyState} />
        ) : (
          recentSearches.map(
            ({ albumId, trackId, image, name, artistName }) => (
              <Link
                onClick={() => setClickedTrackAction(trackId)}
                key={trackId || albumId}
                to={`/album/${albumId}`}
              >
                <ds-image-label
                  data-testid="album-track-card"
                  imageSrc={image}
                  primaryLabel={name}
                  secondaryLabel={artistName}
                />
              </Link>
            )
          )
        )}
      </ds-fieldset>
      {!!recentSearches?.length && (
        <ds-button
          onClick={() => {
            setRecentSearches([])
            clearLocalStorage()
          }}
        >
          Limpar hist√≥rico
        </ds-button>
      )}
    </div>
  )
}

const mapDispatchToProps = {
  setClickedTrackAction: setClickedTrack
}

Home.propTypes = {
  setClickedTrackAction: PropTypes.func
}

export default connect(null, mapDispatchToProps)(Home)
