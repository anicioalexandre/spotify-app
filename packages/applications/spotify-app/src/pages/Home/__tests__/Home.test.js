import React from 'react'
import { screen } from '@testing-library/dom'

import renderWithRedux from '../../../helpers/renderWithRedux'
import renderWithRouter from '../../../helpers/renderWithRouter'
import { recentSearchesMock } from '../__mocks__/recentSearchesMock'
import Home from '..'

const mockHistoryPush = jest.fn()

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useHistory: () => ({ push: mockHistoryPush })
}))

const localStorageMock = (function () {
  let storage = recentSearchesMock
  return {
    getItem: function (key) {
      return JSON.stringify(storage[key])
    },
    setItem: function (key, value) {
      storage[key] = value.toString()
    },
    removeItem: function (key) {
      delete storage[key]
    },
    clear: function () {
      storage = {}
    }
  }
})()
Object.defineProperty(window, 'localStorage', { value: localStorageMock })

describe('Tests for Home page', () => {
  it('renders Home page component', async () => {
    const component = renderWithRedux(renderWithRouter(<Home />))
    expect(component).toMatchSnapshot()
  })
  it('renders three cards as an result of the localStorage content', async () => {
    const { container } = renderWithRedux(renderWithRouter(<Home />))
    expect(container.querySelectorAll('a')).toHaveLength(3)
  })
  it('renders the first card correctly', async () => {
    renderWithRedux(renderWithRouter(<Home />))
    const [firstAlbumCard] = await screen.findAllByTestId('album-track-card')

    const { name, image, artistName } = recentSearchesMock.last_results[0]

    expect(firstAlbumCard.primaryLabel).toBe(name)
    expect(firstAlbumCard.secondaryLabel).toBe(artistName)
    expect(firstAlbumCard.imageSrc).toBe(image)
  })
  it('renders the third card correctly', async () => {
    renderWithRedux(renderWithRouter(<Home />))
    const [, , thirdTrackCard] = await screen.findAllByTestId(
      'album-track-card'
    )

    const { name, image, artistName } = recentSearchesMock.last_results[2]

    expect(thirdTrackCard.primaryLabel).toBe(name)
    expect(thirdTrackCard.secondaryLabel).toBe(artistName)
    expect(thirdTrackCard.imageSrc).toBe(image)
  })
  it('have calls a route change inside the link', async () => {
    const { container } = renderWithRedux(renderWithRouter(<Home />))

    const firstCardHref = container.querySelector('a').href
    const { albumId } = recentSearchesMock.last_results[0]
    expect(firstCardHref).toMatch(albumId)
  })
})
