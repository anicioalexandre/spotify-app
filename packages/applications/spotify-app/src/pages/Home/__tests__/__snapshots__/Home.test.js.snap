// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for Home page renders Home page component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="home-container"
      >
        <ds-input
          label="Busque por álbuns ou músicas"
          placeholder="Comece a escrever..."
        />
        <ds-fieldset
          label="Músicas ou álbuns buscados recentemente"
        >
          <a
            href="/album/0PT5m6hwPRrpBwIHVnvbFX"
          >
            <ds-image-label
              data-testid="album-track-card"
              imagesrc="https://i.scdn.co/image/ab67616d00001e02e3e3b64cea45265469d4cafa"
              primarylabel="Yesterday - Remastered 2009"
              secondarylabel="The Beatles"
            />
          </a>
          <a
            href="/album/1vANZV20H5B4Fk6yf7Ot9a"
          >
            <ds-image-label
              data-testid="album-track-card"
              imagesrc="https://i.scdn.co/image/ab67616d00001e0255612ece447bec5d62c68375"
              primarylabel="Beatles For Sale (Remastered)"
              secondarylabel="The Beatles"
            />
          </a>
          <a
            href="/album/2zeWAYgh0WfMhiHwvUV15b"
          >
            <ds-image-label
              data-testid="album-track-card"
              imagesrc="https://i.scdn.co/image/ab67616d00001e029f08c2a93ef090a1e1e6bf0c"
              primarylabel="Nó na Orelha"
              secondarylabel="Criolo"
            />
          </a>
        </ds-fieldset>
        <ds-button>
          Limpar histórico
        </ds-button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="home-container"
    >
      <ds-input
        label="Busque por álbuns ou músicas"
        placeholder="Comece a escrever..."
      />
      <ds-fieldset
        label="Músicas ou álbuns buscados recentemente"
      >
        <a
          href="/album/0PT5m6hwPRrpBwIHVnvbFX"
        >
          <ds-image-label
            data-testid="album-track-card"
            imagesrc="https://i.scdn.co/image/ab67616d00001e02e3e3b64cea45265469d4cafa"
            primarylabel="Yesterday - Remastered 2009"
            secondarylabel="The Beatles"
          />
        </a>
        <a
          href="/album/1vANZV20H5B4Fk6yf7Ot9a"
        >
          <ds-image-label
            data-testid="album-track-card"
            imagesrc="https://i.scdn.co/image/ab67616d00001e0255612ece447bec5d62c68375"
            primarylabel="Beatles For Sale (Remastered)"
            secondarylabel="The Beatles"
          />
        </a>
        <a
          href="/album/2zeWAYgh0WfMhiHwvUV15b"
        >
          <ds-image-label
            data-testid="album-track-card"
            imagesrc="https://i.scdn.co/image/ab67616d00001e029f08c2a93ef090a1e1e6bf0c"
            primarylabel="Nó na Orelha"
            secondarylabel="Criolo"
          />
        </a>
      </ds-fieldset>
      <ds-button>
        Limpar histórico
      </ds-button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": Object {
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
    Symbol(observable): [Function],
  },
  "unmount": [Function],
}
`;
